data/src/main/java/com/linkedin/data/DataParser.java

```diff
@@ -0,0 +1,121 @@
+/*
+   Copyright (c) 2020 LinkedIn Corp.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.data;
+
+import java.io.IOException;
+
+
+/**
+ * Data parser interface invoked by stream decoder.
+ *
+ * This interface contains methods that are invoked when parsing a Data object.
+ * Each method represents a different kind of event/read action
+ *
+ * Methods can throw IOException as a checked exception to
+ * indicate parsing error.
+ *
+ * @author amgupta1
+ */
+public interface DataParser
+{
+  /**
+   * Internal tokens, used to identify types of elements in data during decoding
+   */
+  enum Token
+  {
+    START_OBJECT,
+    END_OBJECT,
+    START_ARRAY,
+    END_ARRAY,
+    STRING,
```

Comment:
I am not sure this is the best way to model it. Protobuf for example has the notion of several more internal types, like say string reference. An Ordinal model may work better since it supports extensibility.

Reply:
I think this is okay, concrete parser implementations can use internal tokens and then map them to these tokens in `nextToken()`.
the Json parser is already doing this for number types. 

The tokens defined here are the types needed/allowed for DataComplex objects.

Reply:
Yes I had ordinal in mind but these token should be based on datatypes we allow ie. Data.DATA_TYPES. We can wrap specific ordinals to these token like StringRef to String.